### PR TITLE
Feature/555 Add join date to organization members, refactor Organizations controller

### PR DESCRIPTION
**Pull request by tinahollygb**
### Features and Changes

These changes add the date the user joined the organization to the Team members table. It also refactors the organizations routing on the backend. See below for details of each.

#### Refactor of Organizations routing

I have begun introducing router-based routes (as discussed and partially implemented [in this PR](https://github.com/growthbook/growthbook/pull/520)).

This involves the following changes:

- I moved the methods in the organization controller to a separate router.
- I moved all endpoints that start with `/user` to a new users router.
- The `wrapController` function that wraps async error handling around each controller method has been moved as well to the routers directory.
- I moved the 2nd leg of the file uploading to a separate uploads router—the first leg is tightly coupled to discussions and refactoring that is out of scope for this work, so I only moved the 2nd part out of organizations into uploads.

#### Added `dateCreated` to Organization member

I've added the `dateCreated` property to the organization member, which is applied when the organization is created (for the first member), and when the invite is accepted (for all subsequent members).

You'll notice the existing values are blank. We do not have enough information in the database to infer when the user joined the organization and so it was explicitly left blank. 

- Closes https://github.com/growthbook/growthbook/issues/555

### Dependencies

n/a

#### Teams list

As an admin:

1. Invite a new user
2. Use the invite link to sign up as a new user in an incognito browser
3. Go back to the original browser and see the new, joined user with a date and time displayed

#### Areas affected by the refactor

- Upload a file to a discussion and ensure it works
- Perform user actions like updating your name

### Screenshots

Screen Shot 2022-11-03 at 2 44 34 PM

Screen Shot 2022-11-03 at 2 44 48 PM

### CODE REVIEW

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 1]
```diff
@@ -70,70 +62,12 @@ import {
...
   getUnredactedSecretKey,
-} from "../models/ApiKeyModel";
+} from "../../models/ApiKeyModel";
 import {
   accountFeatures,
   getAccountPlan,
   getRoles,
-} from "../util/organization.util";
-
-export async function getUser(req: AuthRequest, res: Response) {
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:80]: Moved to users router.

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 2]
```diff
@@ -188,26 +122,6 @@ export async function getDefinitions(req: AuthRequest, res: Response) {
   });
 }
 
-export async function getUsers(req: AuthRequest, res: Response) {
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:191]: Unused endpoint was deleted

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 3]
```diff
@@ -249,92 +163,6 @@ export async function getActivityFeed(req: AuthRequest, res: Response) {
   }
 }
 
-export async function postWatchItem(
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:252]: Moved to users router.

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 4]
```diff
@@ -249,92 +163,6 @@ export async function getActivityFeed(req: AuthRequest, res: Response) {
...
-  } else {
-    await ensureWatching(userId, org.id, id, "experiments");
-  }
-
-  return res.status(200).json({
-    status: 200,
-  });
-}
-
-export async function postUnwatchItem(
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:286]: Moved to users router.

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 5]
```diff
@@ -249,92 +163,6 @@ export async function getActivityFeed(req: AuthRequest, res: Response) {
...
-    });
-  } catch (e) {
-    res.status(400).json({
-      status: 400,
-      message: e.message,
-    });
-  }
-}
-
-export async function getWatchedItems(req: AuthRequest, res: Response) {
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:318]: Moved to users router.

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 6]
```diff
@@ -368,35 +196,6 @@ export async function getHistory(
   });
 }
 
-export async function putUserName(
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:371]: Moved to users router.

File: packages/back-end/src/routers/organizations/organizations.controller.ts [hunk 7]
```diff
@@ -1270,15 +1069,6 @@ export async function postImportConfig(
   });
 }
 
-export async function putUpload(req: Request, res: Response) {
```
Comment (tinahollygb) [inline packages/back-end/src/routers/organizations/organizations.controller.ts:1273]: Moved to upload router.

File: packages/front-end/components/Settings/Team/MemberList.tsx [hunk 8]
```diff
@@ -84,6 +87,7 @@ const MemberList: FC<{
               <tr key={id}>
                 <td>{member.name}</td>
                 <td>{member.email}</td>
+                <td>{member.dateCreated && datetime(member.dateCreated)}</td>
```
Comment (jdorn) [inline packages/front-end/components/Settings/Team/MemberList.tsx:90]: I think we can figure out the member join date with 99.9% accuracy from existing data.  In the vast majority of cases, the user registers (i.e. creates a document in the `users` collection) immediately before accepting an invitation or creating an organization.

The `users` collection in Mongo doesn't have a `dateCreated` field unfortunately, but it does have the standard Mongo `_id` field, which contains a timestamp.

Given a `UserDocument` object, you can get a timestamp with:

```ts
user._id.getTimestamp()
```

There will be a handful of edge cases where the `UserDocument` date is way before their actual organization join date, but it's  going to be extremely rare.  I think it's better to show the wrong date for 0.1% of users than to show no date for 99.9% of users.

**Review by jdorn — CHANGES_REQUESTED** [review]

File: packages/front-end/components/Settings/Team/MemberList.tsx (hunk 8 above)
Comment (tinahollygb) [reply packages/front-end/components/Settings/Team/MemberList.tsx:90]: I've made this change. 
image

File: packages/back-end/types/organization.d.ts [hunk 9]
```diff
@@ -36,6 +36,7 @@ export type CommercialFeature = "sso" | "advanced-permissions";
 export type CommercialFeaturesMap = Record<AccountPlan, Set<CommercialFeature>>;
 
 export interface MemberRoleInfo {
+  dateCreated?: Date;
```
[the remark this answers is not in the document]
Comment (tinahollygb) [reply packages/back-end/types/organization.d.ts]: I've moved the date to `Member`.

**Review by jdorn — APPROVED** [review]

**Pull request merged**